File: csharp-workshop/Part3.MocksClean/Service/OrderService.cs
using Part3.MocksClean.Domain;
using Part3.MocksClean.Repository;

namespace Part3.MocksClean.Service;

public class OrderService
{
    private readonly IOrderRepository _orderRepository;
    private readonly IStockService _stockService;
    private readonly PricingService _pricingService;
    private readonly IPaymentService _paymentService;
    private readonly IShippingService _shippingService;

    public OrderService(
        IOrderRepository orderRepository,
        IStockService stockService,
        PricingService pricingService,
        IPaymentService paymentService,
        IShippingService shippingService)
    {
        _orderRepository = orderRepository;
        _stockService = stockService;
        _pricingService = pricingService;
        _paymentService = paymentService;
        _shippingService = shippingService;
    }

    public OrderResult PlaceOrder(Order order)
    {
        if (!IsValid(order))
        {
            return new OrderResult.Failure(order.Reject(), "Invalid order");
        }

        var reservations = _stockService.ReserveStock(order.Items);
        if (reservations.Any(r => !r.Reserved))
        {
            return new OrderResult.Failure(order.Reject(), "Insufficient stock");
        }

        double totalPrice = _pricingService.CalculateTotal(order.Items);
        var paymentResult = _paymentService.ProcessPayment(order.Id, order.CustomerId, totalPrice);

        if (paymentResult.Status != PaymentStatus.Success)
        {
            _stockService.ReleaseStock(reservations);
            return new OrderResult.Failure(order.Reject(), "Payment failed");
        }

        var shippingConfirmation = _shippingService.CreateShipment(order);
        var confirmedOrder = order.Confirm(totalPrice);
        _orderRepository.Save(confirmedOrder);
        return new OrderResult.Success(confirmedOrder, shippingConfirmation);
    }

    private static bool IsValid(Order order)
    {
        return order.Status == OrderStatus.Pending
            && order.Items != null
            && order.Items.Count > 0;
    }
}

File: typescript-workshop/part-3-mocks-clean/src/service/OrderService.ts
import { Order } from '../domain/Order';
import { OrderStatus } from '../domain/OrderStatus';
import { PaymentStatus } from '../domain/PaymentStatus';
import { OrderRepository } from '../repository/OrderRepository';
import { OrderResult } from './OrderResult';
import { PaymentService } from './PaymentService';
import { PricingService } from './PricingService';
import { ShippingService } from './ShippingService';
import { StockService } from './StockService';

export class OrderService {
  constructor(
    private readonly orderRepository: OrderRepository,
    private readonly stockService: StockService,
    private readonly pricingService: PricingService,
    private readonly paymentService: PaymentService,
    private readonly shippingService: ShippingService,
  ) {}

  placeOrder(order: Order): OrderResult {
    if (!this.isValid(order)) {
      return OrderResult.failure(order.reject(), 'Invalid order');
    }

    const reservations = this.stockService.reserveStock(order.items);
    if (reservations.some((r) => !r.reserved)) {
      return OrderResult.failure(order.reject(), 'Insufficient stock');
    }

    const totalPrice = this.pricingService.calculateTotal(order.items);
    const paymentResult = this.paymentService.processPayment(order.id, order.customerId, totalPrice);

    if (paymentResult.status !== PaymentStatus.Success) {
      this.stockService.releaseStock(reservations);
      return OrderResult.failure(order.reject(), 'Payment failed');
    }

    const shippingConfirmation = this.shippingService.createShipment(order);
    const confirmedOrder = order.confirm(totalPrice);
    this.orderRepository.save(confirmedOrder);
    return OrderResult.success(confirmedOrder, shippingConfirmation);
  }

  private isValid(order: Order): boolean {
    return order.status === OrderStatus.Pending && order.items.length > 0;
  }
}
